### PR TITLE
MACOS/IPHONE: Sign whole bundle

### PR DESCRIPTION
**Pull request by lephilousophe**
Add signing to MacOS for arm64 builds
This has not been tested but it adds CodeResources file which seem to indicate the whole bundle will be signed.

This does the same thing for iPhone even though I suppose jailbroken devices are less sensitive to the missing signature on all binaries.

### CODE REVIEW

File: dists/ios7/Info.plist.in [hunk 1]
```diff
@@ -11,7 +11,7 @@
 	<key>CFBundleIcons~ipad</key>
 	<dict/>
 	<key>CFBundleIdentifier</key>
-	<string>$(PRODUCT_BUNDLE_IDENTIFIER)</string>
+	<string>ScummVM</string>
```
Comment (criezy) [inline dists/ios7/Info.plist.in:14]: Why this change?
Also it is not consistent with the change made in `Info.plist`. In one you replaced the `${EXECUTABLE_NAME}` and kept the `${PRODUCT_BUNDLE_IDENTIFIER}`, and in the other you did the reverse.

File: ports.mk [hunk 2]
```diff
@@ -182,6 +182,7 @@ ifdef USE_DOCKTILEPLUGIN
 	mkdir -p $(bundle_name)/Contents/PlugIns
 	cp -r scummvm.docktileplugin $(bundle_name)/Contents/PlugIns/
 endif
+	ldid -S $(bundle_name)/Contents
```
Comment (criezy) [inline ports.mk:185]: What is `ldid`? There is no such command on macOS.
I see it was already present for the iOS bundle in `ports.mk`, but as I never compiled the iOS bundle on the command line (I always use Xcode), I never noticed. But adding it to the macOS bundle target will break command line compilation for all those compiling ScummVM on the command line on macOS (including myself). I don't think this is a good idea.

Since the PR title says `sign whole bundle` I assume this is related to signing? The command line utility on macOS to sign application is called `codesign`.

But I am not sure signing (and notarization, which may also be needed) are possible on Linux. This may be relevant: https://github.com/tpoechtrager/osxcross/issues/198

**Review by criezy — CHANGES_REQUESTED** [review]

File: ports.mk (hunk 2 above)
Comment (lephilousophe) [reply ports.mk:185]: I am really a noob with all of this and I am not sure about what is really done by these tools.

I don't know if the ios7 part modifications are needed nor good. I don't even know if the packages produced by buildbot can run on an iPhone (jailbroken or not).
According to what I read, it seems that when a bundle is distributed on Mac, it should be signed as a whole. Up to now, only the binary was signed (either by cctools-port which calls ldid or explicitly by calling ldid directly).
The commit directs ldid to the folder containing the whole bundle to have "manifest" generated and its signature bundled in main binary (if I understood correctly).

`ldid` is a tool to edit entitlements(?) in Mac binaries. It also generates the hashes to sign binaries and can fake(?) sign binaries.
It seems to be used by jailbreak community and it's the only tool available on linux to do so. `codesign` isn't open sourced by Apple and nobody created a mock.
I found another tool called `zsign` while googling but never found a clear explanation about what it does more than `ldid`.
As `ldid` was already used in the Makefile, I thought it was OK. The only change here is to sign the whole directory instead of just the binary.
About that, the Info.plist needs to be changed because `ldid` doesn't handle the `${EXECUTABLE_NAME}` variable and needs the real file name.
It's already like that in dists/iphone/Info.plist but I don't know why.

File: dists/ios7/Info.plist.in (hunk 1 above)
Comment (lephilousophe) [reply dists/ios7/Info.plist.in]: This was a too fast cut and paste.

**Comment by lephilousophe** [comment]
Looking forward to merge this.
@criezy do you have the Apple commands used by Xcode to have a properly working package (on iPhone and on M1)?
I could define a variable with this command on Makefile and override it from environment when using cross compilation tools.

**Comment by criezy** [comment]
Do you have some sources about ldid's ability to sign macOS applications? All I found was related to iOS.

I got `ldid` through MacPorts and played a bit with it. While it can apparently sign the macOS application (no errors, and the executable gets modified), Gatekeeper still doesn't let me open the application (“ScummVM” cannot be opened because the developer cannot be verified.").

My understanding is that the only way to sign a macOS application that will make Gatekeeper happy is to have a paid Apple developer account so that we can get a certificate from Apple and use it for code signing. We might also need notarization.

I am not an expert, but the codesigning command would look like:
`codesign -f -o runtime --deep -s "Certificate Name" ScummVM.app`

And the notarisation would use `altool`.

But until we get an apple developer account for the org we can't use any of this.

**Comment by lephilousophe** [comment]
ldid -S alone fakes signatures.
If I understand well, you need to use -K switch to provide a private key (p12 file which contains associated certificate) to sign binary properly.
I suppose the certificate associated to the private key must be signed by Apple to have Gatekeeper happy.

**Comment by lephilousophe** [comment]
Finally found time to work on this.
We now only expect codesign program. It is supposed to work on MacOS.

For the buildbot, I have a patched ldid to make it work and a shim to mimic codesign behavior.
I am currently building Docker images for it.

**Comment by lephilousophe** [comment]
As there are no comments, I plan to marge this tomorrow.
